gbuiltins/bls381/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+pub const REQUEST_MULTI_SCALAR_MULTIPLICATION_G1: u8 = 2;
+/// Constant defines codec index of [`Request::MultiScalarMultiplicationG2`].
+pub const REQUEST_MULTI_SCALAR_MULTIPLICATION_G2: u8 = 3;
+/// Constant defines codec index of [`Request::ProjectiveMultiplicationG1`].
+pub const REQUEST_PROJECTIVE_MULTIPLICATION_G1: u8 = 4;
+/// Constant defines codec index of [`Request::ProjectiveMultiplicationG2`].
+pub const REQUEST_PROJECTIVE_MULTIPLICATION_G2: u8 = 5;
+
+/// Type that should be used to create a message to the bls12_381 builtin actor.
+#[derive(Encode, Clone, PartialEq, Eq, Debug)]
+#[codec(crate = codec)]
+pub enum Request {
+    /// Request to pairing multi Miller loop for *BLS12-381*.
+    ///
+    /// Encoded:
+    ///   - `a`: `ArkScale<Vec<G1Affine>>`.
+    ///   - `b`: `ArkScale<Vec<G2Affine>>`.
+    #[codec(index = 0)]
+    MultiMillerLoop { a: Vec<u8>, b: Vec<u8> },
+
+    /// Request to pairing final exponentiation for *BLS12-381*.
+    ///
+    /// Encoded: `ArkScale<<Bls12_381::TargetField>`.
+    #[codec(index = 1)]
+    FinalExponentiation { f: Vec<u8> },
+
+    /// Request to multi scalar multiplication on *G1* for *BLS12-381*
+    ///
+    /// Encoded:
+    ///   - `bases`: `ArkScale<Vec<G1Affine>>`.
+    ///   - `scalars`: `ArkScale<Vec<G1Config::ScalarField>>`.
+    #[codec(index = 2)]
+    MultiScalarMultiplicationG1 { bases: Vec<u8>, scalars: Vec<u8> },
+
+    /// Request to multi scalar multiplication on *G2* for *BLS12-381*
+    ///
+    /// Encoded:
+    ///   - `bases`: `ArkScale<Vec<G2Affine>>`.
+    ///   - `scalars`: `ArkScale<Vec<G2Config::ScalarField>>`.
+    #[codec(index = 3)]
+    MultiScalarMultiplicationG2 { bases: Vec<u8>, scalars: Vec<u8> },
+
+    /// Request to projective multiplication on *G1* for *BLS12-381*.
+    ///
+    /// Encoded:
+    ///   - `base`: `ArkScaleProjective<G1Projective>`.
+    ///   - `scalar`: `ArkScale<Vec<u64>>`.
+    #[codec(index = 4)]
+    ProjectiveMultiplicationG1 { base: Vec<u8>, scalar: Vec<u8> },
+
+    /// Request to projective multiplication on *G2* for *BLS12-381*.
+    ///
+    /// Encoded:
+    ///   - `base`: `ArkScaleProjective<G2Projective>`.
+    ///   - `scalar`: `ArkScale<Vec<u64>>`.
+    #[codec(index = 5)]
+    ProjectiveMultiplicationG2 { base: Vec<u8>, scalar: Vec<u8> },
+}
+
+/// The enumeration represents possible common errors for all requests.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[codec(crate = codec)]
+pub enum Error {
+    /// Failed to scale-decode the length of a `Vec<u8>`.
+    DecodeVecLength,
+    /// Failed to scale-decode bytes.
+    DecodeVecData,
+    /// Failed to decode the length of a `Vec<G(1,2)Affine>`.
+    DecodeItemCount,
+}
+
+/// The enumeration contains result to a request.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, derive_more::From)]
+#[codec(crate = codec)]
+pub enum Response {
+    /// Common error.
+    #[from]
+    Error(Error),
```

Comment:
error shouldn't be in Response
builtin framework supports -> Result<YOUR_OK_BYTES, Error> where Error includes your custom variant

Reply:
discussed: use BuiltinError::Custom with the error description